### PR TITLE
fraud: rework unmarshaler registration

### LINKED ISSUE #823: header|fraud: a simple trick to lower allocations with PubSub
Semantically, we have FraudSub and HeaderSub, which rely on libp2p's PubSub. Every msgs in pubsub goes through validation and then to subscription. To validate a msg, we deserialize it, and per each subscription, we deserialize it. In the HeaderSub we even deserialize additionally for msgID gen. With the trick we can do one only and here is it:
PubSub's msg has an internal field [`ValidatorData`](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L223) which we can set during validation. This field is propagated to each subscription that can access it, convert to the correct type, and pass them to the user.

### PR DESCRIPTION
**Pull request by vgonkivs**
Resolves https://github.com/celestiaorg/celestia-node/issues/941

Partly resolves #823

### CODE REVIEW

**Comment by vgonkivs** [comment]
Swamp test that was failed is not affected by these changes:

```
2022-07-23T10:49:31.7278404Z --- FAIL: TestSyncFullWithBridge (0.05s)
2022-07-23T10:49:31.7299700Z panic: interface conversion: types.TMEventData is nil, not types.EventDataNewBlock [recovered]
2022-07-23T10:49:31.7300908Z 	panic: interface conversion: types.TMEventData is nil, not types.EventDataNewBlock
2022-07-23T10:49:31.7301187Z 
2022-07-23T10:49:31.7301307Z goroutine 417508 [running]:
2022-07-23T10:49:31.7301604Z testing.tRunner.func1.2({0x1e59be0, 0xc014722240})
2022-07-23T10:49:31.7302013Z 	/opt/hostedtoolcache/go/1.18.4/x64/src/testing/testing.go:1389 +0x24e
2022-07-23T10:49:31.7303124Z testing.tRunner.func1()
2022-07-23T10:49:31.7303971Z 	/opt/hostedtoolcache/go/1.18.4/x64/src/testing/testing.go:1392 +0x39f
2022-07-23T10:49:31.7304273Z panic({0x1e59be0, 0xc014722240})
2022-07-23T10:49:31.7304625Z 	/opt/hostedtoolcache/go/1.18.4/x64/src/runtime/panic.go:838 +0x207
2022-07-23T10:49:31.7305555Z github.com/celestiaorg/celestia-node/node/tests/swamp.(*Swamp).WaitTillHeight(0xc00d1a4000, {0x2dd24e8, 0xc003ef05a0}, 0x14)
2022-07-23T10:49:31.7306199Z 	/home/runner/work/celestia-node/celestia-node/node/tests/swamp/swamp.go:133 +0x3ca
2022-07-23T10:49:31.7306803Z github.com/celestiaorg/celestia-node/node/tests.TestSyncFullWithBridge(0xc01a63c9c0)
2022-07-23T10:49:31.7309332Z 	/home/runner/work/celestia-node/celestia-node/node/tests/sync_test.go:138 +0xae
2022-07-23T10:49:31.7309952Z testing.tRunner(0xc01a63c9c0, 0x2a66950)
2022-07-23T10:49:31.7314298Z 	/opt/hostedtoolcache/go/1.18.4/x64/src/testing/testing.go:1439 +0x102
2022-07-23T10:49:31.7314809Z created by testing.(*T).Run
2022-07-23T10:49:31.7315173Z 	/opt/hostedtoolcache/go/1.18.4/x64/src/testing/testing.go:1486 +0x35f
2022-07-23T10:49:31.7787821Z FAIL	github.com/celestiaorg/celestia-node/node/tests	66.793s
```

**Comment by vgonkivs** [comment]
In a private discussion with @Wondertan we agreed to return `fraud.Subscription` back in order to avoid hard bindings on `pubsub.Subscription`

File: fraud/registry.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package fraud
+
+var DefaultUnmarshalers = map[ProofType]ProofUnmarshaler{}
```
Comment (Wondertan) [inline fraud/registry.go:3]: It's better to be protected with a lock and thus encapsulated behind a getter, like Register is for writes

Comment (vgonkivs) [reply fraud/registry.go]: done

File: fraud/registry.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	Register(BadEncoding, UnmarshalBEFP)
+}
+
+func Register(proofType ProofType, unmarshaler ProofUnmarshaler) {
+	unmarshalersLk.Lock()
+	unmarshalersLk.Unlock()
+	defaultUnmarshalers[proofType] = unmarshaler
+}
+
+func GetUnmarshaler(proofType ProofType) (ProofUnmarshaler, error) {
```
Comment (Wondertan) [inline fraud/registry.go:21]: Docs?

Comment (Wondertan) [inline fraud/registry.go:21]: What about `Unmarshal` func that takes both type and method> There is no real usage for an `unmarhaller` besides calling it, so we can call it right away.

File: fraud/registry.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	"fmt"
+	"sync"
+)
+
+var unmarshalersLk = sync.RWMutex{}
+var defaultUnmarshalers = map[ProofType]ProofUnmarshaler{}
+
+func init() {
+	Register(BadEncoding, UnmarshalBEFP)
+}
```
Comment (Wondertan) [inline fraud/registry.go:13]: Let's move it close to the definition of the BEFP itself. Technically, now, BEFP can now live outside of this package - in DASer :) (but that does not matter too much)

**Review by Wondertan — COMMENTED** [review]
We should have done this from the beginning; however, I convinced myself that this is not a good approach because of the global state for some weird reason.

P.S. Would be nice to include here [this trick](https://github.com/celestiaorg/celestia-node/issues/823)

File: fraud/registry.go (hunk 3 above)
Comment (vgonkivs) [reply fraud/registry.go:13]: I suggest to keep it behind ` fraud.NewService`. wdyt? @Wondertan

File: fraud/registry.go (hunk 2 above)
Comment (vgonkivs) [reply fraud/registry.go:21]: Yes, will add it as a separate commit

**Comment by vgonkivs** [comment]
> P.S. Would be nice to include here https://github.com/celestiaorg/celestia-node/issues/823
Why not, will do this 👍

File: fraud/registry.go (hunk 2 above)
Comment (vgonkivs) [reply fraud/registry.go:21]: May be let's do `func Unmarshal(p proofType, data []byte) (Proof, error)` then? 

> func that takes both type and method

We should somehow get this method. I don't like calling a global mutex inside `fraud.Service` methods.

File: fraud/service.go [hunk 4]
```diff
@@ -37,60 +40,30 @@ func NewService(p *pubsub.PubSub, getter headerFetcher, ds datastore.Datastore)
...
-	return nil
-}
-
-func (f *service) UnregisterUnmarshaler(proofType ProofType) error {
-	f.topicsLk.Lock()
-	defer f.topicsLk.Unlock()
-	t, ok := f.topics[proofType]
-	if !ok {
-		return fmt.Errorf("fraud: unmarshaler for %s proof is not registered", proofType)
+	if err != nil {
```
Comment (Wondertan) [inline fraud/service.go:62]: Why unlock in between?

File: fraud/registry.go (hunk 2 above)
Comment (Wondertan) [reply fraud/registry.go]: LgTM

File: fraud/service.go [hunk 5]
```diff
@@ -19,11 +18,15 @@ const (
 	fetchHeaderTimeout = time.Minute * 2
 )
 
+func init() {
+	Register(BadEncoding, UnmarshalBEFP)
+}
```
Comment (Wondertan) [inline fraud/service.go:23]: Its fine to register it next to the service considering that BEFP is defined in the same pkg. However, current model allows anyone outside of the fraud pkg to register their own Fraud, so in that case the registering is better be close to the definition of the fraud type and maybe we should follow that approach.

File: fraud/subscription.go [hunk 6]
```diff
@@ -2,57 +2,33 @@ package fraud
...
-		return nil, err
-	}
-
-	return &subscription{sub, t.unmarshal}, nil
 }
 
 func (s *subscription) Proof(ctx context.Context) (Proof, error) {
+	if s.subscription == nil {
+		return nil, errors.New("fraud: subscription is not created")
+	}
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply fraud/subscription.go:20]: Just to be consistent... Can also panic here.

File: fraud/helpers.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+func join(p *pubsub.PubSub, proofType ProofType,
+	validate func(context.Context, ProofType, *pubsub.Message) pubsub.ValidationResult) (*pubsub.Topic, error) {
+	t, err := p.Join(getSubTopic(proofType))
+	if err != nil {
+		return nil, err
+	}
+	err = p.RegisterTopicValidator(
+		getSubTopic(proofType),
+		func(ctx context.Context, _ peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply fraud/helpers.go:24]: Yes, that makes sense. The only question is - will re-gossipper send message that didn't pass validation. I mean, processIncoming will also detect Fraud Proof and reject this message at all, no?

File: fraud/subscription.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Wondertan) [reply fraud/subscription.go:20]: To be consistent with what?

File: fraud/helpers.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Wondertan) [reply fraud/helpers.go:24]: If validation is rejected, then nothing is reoggiped

[the remark this answers is not in the document]
Comment (Wondertan) [reply fraud/helpers.go:24]: But we cannot trust the regossiper on the other side as code can be modified

File: fraud/subscription.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vgonkivs) [reply fraud/subscription.go:20]: we have a pointer inside struct. Let's check it 😄

File: fraud/helpers.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vgonkivs) [reply fraud/helpers.go:24]: will be fixed in #954

File: fraud/service.go (hunk 5 above)
Comment (vgonkivs) [reply fraud/service.go]: moved to Daser

File: fraud/proof.go [hunk 8]
```diff
@@ -15,6 +16,14 @@ func (e *ErrFraudExists) Error() string {
 	return fmt.Sprintf("fraud: %s proof exists\n", e.Proof[0].Type())
 }
 
+type ErrNoUnmarshaler struct {
+	proofType ProofType
```
Comment (Wondertan) [inline fraud/proof.go:20]: Public?

**Review by Wondertan — APPROVED** [review]
LGTM. Thanks!

**Comment by Wondertan** [comment]
Next, the blacklisting issue

File: fraud/proof.go (hunk 8 above)
Comment (vgonkivs) [reply fraud/proof.go:20]: good idea. will make it private

File: fraud/proof.go [hunk 9]
```diff
@@ -44,5 +53,38 @@ type Proof interface {
...
+			log.Errorw("reading next proof failed", "err", err)
+		}
+		return
+	}
+
+	handle(proof)
+}
+
+// Unmarshal converts raw bytes into respective Proof type.
+func Unmarshal(proofType ProofType, msg []byte) (Proof, error) {
```
Comment (renaynay) [inline fraud/proof.go:82]: does this have to be exported?

File: fraud/registry.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import (
+	"sync"
+)
+
+var unmarshalersLk = sync.RWMutex{}
+var defaultUnmarshalers = map[ProofType]ProofUnmarshaler{}
+
+// Register sets unmarshaler in map by provided ProofType.
+func Register(proofType ProofType, unmarshaler ProofUnmarshaler) {
```
Comment (renaynay) [inline fraud/registry.go:11]: Does this have to be exported?

File: fraud/subscription.go [hunk 11]
```diff
@@ -2,57 +2,32 @@ package fraud
...
 import (
 	"context"
+	"fmt"
+	"reflect"
 
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
 )
 
-// subscription handles Fraud Proof from the pubsub topic.
+// subscription wraps pubsub subscription and handle Fraud Proof from the pubsub topic.
```
Comment (renaynay) [inline fraud/subscription.go:11]:
```suggestion
// subscription wraps pubsub subscription and handles Fraud Proof from the pubsub topic.
```

File: fraud/service.go [hunk 12]
```diff
@@ -104,24 +71,20 @@ func (f *service) Broadcast(ctx context.Context, p Proof) error {
 	if !ok {
 		return fmt.Errorf("fraud: unmarshaler for %s proof is not registered", p.Type())
 	}
-	return t.publish(ctx, bin)
+	return t.Publish(ctx, bin)
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply fraud/service.go:74]: I removed our internal `topic` struct. Now it's pubSub publish

File: fraud/registry.go (hunk 10 above)
Comment (vgonkivs) [reply fraud/registry.go:11]: Yes, why not? So we can register custom unmarshaler everywhere from code.

Comment (Wondertan) [reply fraud/registry.go:13]: Yes it has

File: fraud/proof.go [hunk 13]
```diff
@@ -15,6 +16,14 @@ func (e *ErrFraudExists) Error() string {
 	return fmt.Sprintf("fraud: %s proof exists\n", e.Proof[0].Type())
 }
 
+type ErrNoUnmarshaler struct {
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply fraud/proof.go]: made private.

File: docs/adr/adr-006-fraud-service.md [hunk 14]
```diff
@@ -177,12 +172,11 @@ In addition, `das.Daser`:
     stores   map[ProofType]datastore.Datastore
 
     topics map[ProofType]*pubsub.Topic
-    unmarshallers map[ProofType]ProofUnmarshaller
+   
+    getter headerFetcher
+	 ds    datastore.Datastore
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply docs/adr/adr-006-fraud-service.md]: fixed

File: fraud/registry.go (hunk 10 above)
Comment (renaynay) [reply fraud/registry.go:13]: ok.

File: fraud/proof.go (hunk 9 above)
Comment (renaynay) [reply fraud/proof.go:82]: ok

File: fraud/service.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (renaynay) [reply fraud/service.go:74]: aha got it

**Review by renaynay — APPROVED** [review]

**Pull request merged**